server.cu

```diff
@@ -18,2040 +18,1809 @@
 
-        return result;
-    }
+    return result;
+}
 
-    case RPC_nvmlDeviceGetVbiosVersion:
-    {
-        nvmlDevice_t device;
-        unsigned int length;
 
-        if (read(connfd, &device, sizeof(nvmlDevice_t)) < 0 ||
-            read(connfd, &length, sizeof(unsigned int)) < 0)
-            return -1;
+int handle_nvmlDeviceRegisterEvents(int connfd) {
+    nvmlDevice_t device;
+    unsigned long long eventTypes;
+    nvmlEventSet_t set;
 
-        char version[length];
-        nvmlReturn_t result = nvmlDeviceGetVbiosVersion(device, version, length);
+    if (read(connfd, &device, sizeof(nvmlDevice_t)) < 0 || read(connfd, &eventTypes, sizeof(unsigned long long)) < 0 || read(connfd, &set, sizeof(nvmlEventSet_t)) < 0)
+        return -1;
 
-        if (write(connfd, version, length) < 0)
-            return -1;
+    nvmlReturn_t result = nvmlDeviceRegisterEvents(device, eventTypes, set);
+    return result;
+}
 
-        return result;
-    }
+int handle_nvmlEventSetCreate(int connfd) {
+    nvmlEventSet_t set;
+    nvmlReturn_t result = nvmlEventSetCreate(&set);
 
-    case RPC_nvmlDeviceGetViolationStatus:
-    {
-        nvmlDevice_t device;
-        nvmlPerfPolicyType_t perfPolicyType;
-        nvmlViolationTime_t violTime;
+    if (write(connfd, &set, sizeof(nvmlEventSet_t)) < 0)
+        return -1;
 
-        if (read(connfd, &device, sizeof(nvmlDevice_t)) < 0 ||
-            read(connfd, &perfPolicyType, sizeof(nvmlPerfPolicyType_t)) < 0)
-            return -1;
+    return result;
+}
 
-        nvmlReturn_t result = nvmlDeviceGetViolationStatus(device, perfPolicyType, &violTime);
+int handle_nvmlEventSetFree(int connfd) {
+    nvmlEventSet_t set;
 
-        if (write(connfd, &violTime, sizeof(nvmlViolationTime_t)) < 0)
-            return -1;
+    if (read(connfd, &set, sizeof(nvmlEventSet_t)) < 0)
+        return -1;
 
-        return result;
-    }
+    return nvmlEventSetFree(set);
+}
 
-    case RPC_nvmlDeviceOnSameBoard:
-    {
-        nvmlDevice_t device1, device2;
-        int onSameBoard;
+int handle_nvmlEventSetWait_v2(int connfd) {
+    nvmlEventSet_t set;
+    nvmlEventData_t data;
+    unsigned int timeoutms;
 
-        if (read(connfd, &device1, sizeof(nvmlDevice_t)) < 0 ||
-            read(connfd, &device2, sizeof(nvmlDevice_t)) < 0)
-            return -1;
+    if (read(connfd, &set, sizeof(nvmlEventSet_t)) < 0 || read(connfd, &timeoutms, sizeof(unsigned int)) < 0)
+        return -1;
 
-        nvmlReturn_t result = nvmlDeviceOnSameBoard(device1, device2, &onSameBoard);
+    nvmlReturn_t result = nvmlEventSetWait_v2(set, &data, timeoutms);
 
-        if (write(connfd, &onSameBoard, sizeof(int)) < 0)
-            return -1;
+    if (write(connfd, &data, sizeof(unsigned long long)) < 0)
+        return -1;
 
-        return result;
-    }
+    return result;
+}
 
-    case RPC_nvmlDeviceValidateInforom:
-    {
-        nvmlDevice_t device;
+static std::unordered_map<unsigned int, RequestHandler> opHandlers = {
```

Comment:
Try making this an array instead? the `#define`'s are all sequential so we should be able to merge them into one bigass array which would be faster than a map.